Function Attributes/README.md

```diff
@@ -0,0 +1,524 @@
+- Behavioral attributes (`pure`, `nothrow`, `@nogc`)
+- Code safety attributes (`@safe`, `@trusted`, `@system`)
+
+## Return type attributes
+
+These attributes are about the return types of the functions.
+
+### `auto`
+
+When a function is annotated with `auto`, its return type doesn't need to be specified because it is automatically inferred by the compiler from the return expression(s).
+Let's take a look at the following example:
+
+```d
+auto func(int i)
+{
+    if (i < 0)
+    {
+        return i;      // returns 'int' here
+    }
+
+    return i * 1.5;    // returns 'double' here
+}
+
+void main()
+{
+    // The return type of the function is 'double'
+    auto result = func(42);
+    static assert(is (typeof(result) == double));
+}
+```
+
+If the `auto` function has only one `return` expression, then the compiler will take the type of that expression as the return type of the function. 
+If there are multiple `return` expressions, the return expression of the function will be the _common type_. 
+In the example above, the common type of `int` and `double` is `double`, and that's because `double` can hold any `int` value, but others as well.
+
+
+### `ref`
+
+If we want to modify a parameter and that change to be visible in the caller, then that parameter should be marked as `ref`.
+
+```d
+void bump(ref int x) { ++x; }
+
+unittest {
+    int x = 1;
+    bump(x);
+    assert(x == 2);
+}
+```
+
+If a function expects a `ref` parameter, then the argument that is passed should be an `lvalue`, otherwise the code won't compile:
+```d
+bump(5) // Error! Cannot bind an rvalue to a ref parameter
+```
+
+The `ref` attribute can be also used for the return type of a function.
+In that case, the return value of the function is an `lvalue` as well.
+
+Let's add `ref` to `bump`.
+```d
+ref int bump(ref int x) { 
+    return ++x;
+}
+
+unittest {
+    int x = 1;
+    bump(bump(x)); // Two increments
+    assert(x == 3);
+}
+``` 
+The inner call to `bump` returns an `lvalue` that is then passed again to `bump` and it works because `bump` expects a `ref`.
+Had the definition of `bump` looked like this:
+```d
+int bump(ref int x) { return ++x; }
+```
+the compiler would have thrown an error for the call `bump(bump(x))`.
+That is because it would have been an attempt to bind an `rvalue` to a function that expects an `lvalue`, that is, a `ref`.
+
+
+### `auto ref`
+
+Similar to `auto`, the return type of an `auto ref` function is deduced by the compiler.
+This is especially useful for templates like the one below:
+
+```d
+auto ref foo(string magic)() 
+{
+   static if(magic == "use ref")
+   {
+     int* x = new int;
+     return *x;
+   } 
+   else
+   {
+     return 0;
+   }
+}
+
+void main() {
+    foo!""(); // returns rvalue, i.e. 0
+    foo!"use ref"(); // returns lvalue, i.e. x, a pointer
+}
+```
+
+Each unique set of compile-time arguments will generate a different function.
+
+### `inout`
```

Comment:
I would delete this section. `inout` is a quirk and many advocate for its removal from the language.

Reply:
ok, i guess we can leave it as is.